host-controller/src/main/java/org/jboss/as/host/controller/mgmt/HostControllerRegistrationHandler.java

```diff
@@ -346,6 +347,13 @@ public void execute(final OperationContext context, final ModelNode operation) t
                     registrationContext.failed(failure, SlaveRegistrationException.ErrorCode.INCOMPATIBLE_VERSION, failure.getMessage());
                     throw failure;
                 }
+                // Ensure feature stream of host is compatible
+                LocalHostControllerInfo domainInfo = domainController.getLocalHostInfo();
+                if (!domainInfo.enables(hostInfo)) {
```

Comment:
The agreement was to allow only paired domains, I mean, STABLE <->, PREVIEW <-> PREVIEW, EXPERIMENTAL <-> EXPERIMENTAL, we don't want to deal with host exclusions with host using mixed streams. However, looking at the FeatureStream.enables(FeatureStream stream), here we are not rejecting paired streams.

I guess we need to be more restrictive here.

Reply:
@yersan Forgive my ignorance in this area, but I'm struggling a bit to recall the reasons why host exclusions were specifically a problem WRT feature streams...

AIUI, host exclusions are meant to handle the case where a subsystem's module referenced within the configuration of the domain controller does not exist on a host controller because it is running an earlier version prior to the module's existence.  Does that sound right?
Do host exclusions serve any other purpose?

If this is the case, I do not see why a host supporting PREVIEW features should not be allowed to connect to a domain controller supporting only STABLE features, since the configuration sent from the domain controller will always be valid on the host (and also why I said the logic above is backwards).  So, host1 using the STABLE feature stream, host2 running the PREVIEW feature stream, and host3 running the EXPERIMENTAL feature stream _should_ all be able to safely join a domain controller so long as it is running the STABLE feature stream - since the feature stream does not affect the modules available to a host (unless we also decided to also allow modules to be associated with a feature stream - but we'll ignore that for now).  Am I missing something?

The complexity, I think, comes when the hosts are of mixed versions - since we now require model transformation to the management API version supported by the host, and we don't want to introduce the complexity/maintenance of supporting model transformations for non-STABLE feature streams.  Therefore, it seems to me that our host registration check would need to be more restrictive than your comment above; i.e. we can allow a domain of hosts running different versions, so long as all hosts use the STABLE stream; or a domain of hosts running the same versions, so long as all hosts are using the same feature stream.  Does that make sense?

Reply:
> Forgive my ignorance in this area, but I'm struggling to recall the reasons why host exclusions were specifically a problem WRT feature streams...

Of course, there is nothing to forgive @pferraro :)

> AIUI, host exclusions are meant to handle the case where a subsystem's module referenced within the configuration of the domain controller does not exist on a host controller because it is running an earlier version prior to the module's existence. Does that sound right?

Yes, that's right. We can talk about extensions, host exclusions exclude extensions.

> Do host exclusions serve any other purpose?

No, just to exclude extensions (with their subsystem) that are unknown by the secondary host. They need to be ignored on the secondary host. When the secondary host gets a copy of the management model from the domain controller, it needs to know which extensions must be ignored based on its kernel version. Sadly we control the list of those extensions on the Domain Controller side, a smarter approach could be host controller to advise what are the extensions known (that's a different topic we don't need to talk here).

> If this is the case, I do not see why a host supporting PREVIEW features should not be allowed to connect to a domain controller supporting only STABLE features, since the configuration sent from the domain controller will always be valid on the host (and also why I said the logic above is backwards). So, host1 using the STABLE feature stream, host2 running the PREVIEW feature stream, and host3 running the EXPERIMENTAL feature stream should all be able to safely join a domain controller so long as it is running the STABLE feature stream - since the feature stream does not affect the modules available to a host (unless we also decided to also allow modules to be associated with a feature stream - but we'll ignore that for now). Am I missing something?

You're right about the above, let's forget mixed domains here. Mixed domain use case is addressed in your third paragraph. I see your reasoning and agree with it; any host controller will support a superset of the extensions sent by the domain controller, so the secondary hosts always be able to understand any possible extension sent by the Domain Controller. We even don't need to worry about host-exlusions for extensions added only on EXPERIMENTAL/PREVIEW. Got it. 

(**) Theoretically, we could have secondary hosts running on a feature stream level above the Domain Controller feature stream.

> The complexity, I think, comes when the hosts are of mixed versions - since we now require model transformation to the management API version supported by the host, and we don't want to introduce the complexity/maintenance of supporting model transformations for non-STABLE feature streams. Therefore, it seems to me that our host registration check would need to be more restrictive than your comment above; i.e. we can allow a domain of hosts running different versions, so long as all hosts use the STABLE stream; or a domain of hosts running the same versions, so long as all hosts are using the same feature stream. Does that make sense?

Right, that makes sense, and indeed Transformers was what I had in mind but I only talked about host exclusions. Yes, this is what we have discussed in the past, they should be running at the same level, at least from the point of view of different HC kernel versions. We will always have (DC <-> HC) in {STABLE <-> STABLE, PREVIEW <-> PREVIEW, EXPERIMENTAL <-> EXPERIMENTAL}, otherwise we reject the registration.

We can go further and do it only if the kernel versions are different, falling back to the discussion (**) but there is no transformers involved, but, there is no gain with that. I would vote for the most restrictive one: paired feature stream versions.

Let me know if I'm missing something.

Reply:
@yersan Thanks for the explanations.  I think I get it now.
I'll update the last commit to reject legacy hosts if the domain controller is not using the STABLE stream.

Reply:
Just for the record, this last step rejecting when not using STABLE for legacy hosts has not been done yet

Reply:
Thanks for the reminder... :)  Will do this today.

Reply:
Looks good, thanks @pferraro , I've just added two minor optional comments related to this but I left them up to your criteria, please resolve this conversation from your side at your convenience.

Reply:
This can be resolved now.